### PR TITLE
Implement first party ephemeral storage support

### PR DESCRIPTION
**Pull request by goodov**
Add support for first party website contexts cleanup using COOKIE_SETTINGS_SESSION_ONLY.

Resolves https://github.com/brave/brave-browser/issues/15906

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/net/base/features.cc [hunk 1]
```diff
@@ -17,5 +17,8 @@ const base::FeatureParam<int> kBraveEphemeralStorageKeepAliveTimeInSeconds = {
     &kBraveEphemeralStorageKeepAlive,
     "BraveEphemeralStorageKeepAliveTimeInSeconds", 30};
 
+const base::Feature kBraveFirstPartyEphemeralStorage{
+    "BraveFirstPartyEphemeralStorage", base::FEATURE_DISABLED_BY_DEFAULT};
```
Comment (iefremov) [inline chromium_src/net/base/features.cc:21]: I guess people would like to have a toggle on brave://flags

File: chromium_src/components/content_settings/core/common/cookie_settings_base.h [hunk 2]
```diff
@@ -10,26 +10,24 @@
...
-  ScopedEphemeralStorageAwareness(ScopedEphemeralStorageAwareness&&);
-  ScopedEphemeralStorageAwareness& operator=(ScopedEphemeralStorageAwareness&&);
-  ~ScopedEphemeralStorageAwareness();
-
- private:
-  base::AutoReset<bool> ephemeral_storage_aware_auto_reset_;
+enum class EphemeralStorageAwareType {
+  kNone,
+  kAware,
+  kNotAwareButAllowIn1pEphemeralMode,
```
Comment (iefremov) [inline chromium_src/components/content_settings/core/common/cookie_settings_base.h:16]: it needs comments

File: browser/ephemeral_storage/ephemeral_storage_1p_browsertest.cc [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        ContentSettingsPattern::FromString(
+            base::StrCat({"[*.]", url.host_piece(), ":*"})),
+        ContentSettingsPattern::Wildcard(), ContentSettingsType::COOKIES,
+        content_setting);
+  }
+
+  // Helper to load easy-to-use Indexed DB API.
+  void LoadIndexedDbHelper(RenderFrameHost* host) {
+    const char kLoadIndexMinScript[] =
+        "new Promise((resolve) => {"
```
Comment (iefremov) [inline browser/ephemeral_storage/ephemeral_storage_1p_browsertest.cc:44]: why not use raw string literal?

File: chromium_src/content/browser/devtools/protocol/network_handler.cc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "content/browser/devtools/protocol/network_handler.h"
+
+#define BRAVE_COOKIE_RETRIEVER_NETWORK_SERVICE_RETRIEVE_ARGS \
+  const url::Origin &top_frame_origin,
+
+#define BRAVE_COOKIE_RETRIEVER_NETWORK_SERVICE_RETRIEVE_BODY \
+  cookie_options.set_top_frame_origin(top_frame_origin);
+
+#define BRAVE_NETWORK_HANDLER_GET_COOKIES_RETREIVE_CALL_ARGS \
```
Comment (iefremov) [inline chromium_src/content/browser/devtools/protocol/network_handler.cc:14]: retrIEve

**Comment by iefremov** [comment]
It's definitely worth adding comments here and there, since this is one of the most complex features and the code is scattered among different overrides

**Comment by iefremov** [comment]
pls check the following scenario:

0. Enable the feature, add `[*.]yandex.ru` to `Always clear cookies when windows are closed`
1. open internet.ya.ru, note the yandexuid cookie value
2. open any site with yandex scripts, e.g. ngs.ru. By default, 3rdparty cookies are blocked
3. Disable shields on ngs.ru, check `yandexuid`. ATM the value differs from 1p cookie set by internet.ya.ru (which is not the case when the feature is off)
4. Also check the cookies view in omnibox -> lock icon -> cookies. In this scenario `blocked` tab has a bunch of `allowed` cookies, which does't look right

**Comment by goodov** [comment]
The current state is now close to production and it's okay to review it.

**Comment by goodov** [comment]
Removed DevTools cookies display support in 1PES mode from this PR. Will add this as a separate thing.

**Review by mariospr — COMMENTED** [review]
Patching/chromium_src-wise LGTM, but I very much agree with @iefremov on that all this code needs comments in several places as it's already quite hard to follow how everything fits together.

**Review by iefremov — APPROVED** [review]

**Comment by iefremov** [comment]
@mariospr PTAL
